File: src/cmd/6g/cplx.c
// Copyright 2009 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

#include "gg.h"

static void	subnode(Node *nr, Node *ni, Node *nc);
static void	negate(Node *n);
static void	zero(Node *n);
static int	isimag1i(Node*);

#define	CASE(a,b)	(((a)<<16)|((b)<<0))

/*
 * generate:
 *	res = n;
 * simplifies and calls gmove.
 * perm is
 *	0 (r,i) -> (r,i)
 *	1 (r,i) -> (-i,r)   *1i
 *	2 (r,i) -> (i,-r)   /1i
 */
void
complexmove(Node *f, Node *t, int perm)
{
	int ft, tt;
	Node n1, n2, n3, n4, nc;

	if(debug['g']) {
		dump("\ncomplex-f", f);
		dump("complex-t", t);
	}

	if(!t->addable)
		fatal("to no addable");

	ft = simsimtype(f->type);
	tt = simsimtype(t->type);
	switch(CASE(ft,tt)) {

	default:
		fatal("complexmove: unknown conversion: %T -> %T\n",
			f->type, t->type);

	case CASE(TCOMPLEX64,TCOMPLEX64):
	case CASE(TCOMPLEX64,TCOMPLEX128):
	case CASE(TCOMPLEX128,TCOMPLEX64):
	case CASE(TCOMPLEX128,TCOMPLEX128):
		// complex to complex move/convert
		// make from addable
		if(!f->addable) {
			tempname(&n1, f->type);
			complexmove(f, &n1, 0);
			f = &n1;
		}

		subnode(&n1, &n2, f);
		subnode(&n3, &n4, t);

		// perform the permutations.
		switch(perm) {
		case 0:	// r,i => r,i
			gmove(&n1, &n3);
			gmove(&n2, &n4);
			break;
		case 1: // r,i => -i,r
			regalloc(&nc, n3.type, N);
			gmove(&n2, &nc);
			negate(&nc);
			gmove(&n1, &n4);
			gmove(&nc, &n3);
			regfree(&nc);
			break;
		case 2: // r,i => i,-r
			regalloc(&nc, n4.type, N);
			gmove(&n1, &nc);
			negate(&nc);
			gmove(&n2, &n3);
			gmove(&nc, &n4);
			regfree(&nc);
			break;
		}
		break;

	case CASE(TFLOAT32,TCOMPLEX64):
	case CASE(TFLOAT32,TCOMPLEX128):
	case CASE(TFLOAT64,TCOMPLEX64):
	case CASE(TFLOAT64,TCOMPLEX128):
		// float to complex goes to real part

		regalloc(&n1, types[ft], N);
		cgen(f, &n1);
		subnode(&n3, &n4, t);

		// perform the permutations.
		switch(perm) {
		case 0:	// no permutations
			gmove(&n1, &n3);
			zero(&n4);
			break;
		case 1:
			gmove(&n1, &n4);
			zero(&n3);
			break;
		case 2:
			negate(&n1);
			gmove(&n1, &n4);
			zero(&n3);
			break;
		}
		regfree(&n1);
		break;
	}
}

int
complexop(Node *n, Node *res)
{
	if(n != N && n->type != T)
	if(iscomplex[n->type->etype]) {
		switch(n->op) {
		case OCONV:
		case OADD:
		case OSUB:
		case OMUL:
		case ODIV:
		case OMINUS:
			goto yes;
		}
//dump("complexop no", n);
	}
	return 0;

yes:
	return 1;
}

void
complexgen(Node *n, Node *res)
{
	Node *nl, *nr;
	Node n1, n2, n3, n4, n5, n6;
	Node ra, rb, rc, rd;
	int tl, tr;

	if(debug['g']) {
		dump("\ncomplex-n", n);
		dump("complex-res", res);
	}

	// perform conversion from n to res
	tl = simsimtype(res->type);
	tl = cplxsubtype(tl);
	tr = simsimtype(n->type);
	tr = cplxsubtype(tr);
	if(tl != tr) {
		if(!n->addable) {
			tempname(&n1, n->type);
			complexmove(n, &n1, 0);
			n = &n1;
		}
		complexmove(n, res, 0);
		return;
	}

	nl = n->left;
	if(nl == N)
		return;
	nr = n->right;

	// make both sides addable in ullman order
	if(nr != N) {
		if(nl->ullman > nr->ullman && !nl->addable) {
			tempname(&n1, nl->type);
			complexgen(nl, &n1);
			nl = &n1;
		}
		if(!nr->addable) {
			tempname(&n2, nr->type);
			complexgen(nr, &n2);
			nr = &n2;
		}
	}
	if(!nl->addable) {
		tempname(&n1, nl->type);
		complexgen(nl, &n1);
		nl = &n1;
	}

	switch(n->op) {
	default:
		fatal("opcode %O", n->op);
		break;

	case OCONV:
		complexmove(nl, res, 0);
		break;

	case OMINUS:
		subnode(&n1, &n2, nl);
		subnode(&n5, &n6, res);

		regalloc(&ra, n5.type, N);
		gmove(&n1, &ra);
		negate(&ra);
		gmove(&ra, &n5);
		regfree(&ra);

		regalloc(&ra, n5.type, N);
		gmove(&n2, &ra);
		negate(&ra);
		gmove(&ra, &n6);
		regfree(&ra);
		break;

	case OADD:
	case OSUB:

		subnode(&n1, &n2, nl);
		subnode(&n3, &n4, nr);
		subnode(&n5, &n6, res);

		regalloc(&ra, n5.type, N);
		gmove(&n1, &ra);
		gins(optoas(n->op, n5.type), &n3, &ra);
		gmove(&ra, &n5);
		regfree(&ra);

		regalloc(&ra, n6.type, N);
		gmove(&n2, &ra);
		gins(optoas(n->op, n6.type), &n4, &ra);
		gmove(&ra, &n6);
		regfree(&ra);
		break;

	case OMUL:
		if(isimag1i(nr)) {
			complexmove(nl, res, 1);
			break;
		}
		if(isimag1i(nl)) {
			complexmove(nr, res, 1);
			break;
		}

		subnode(&n1, &n2, nl);
		subnode(&n3, &n4, nr);
		subnode(&n5, &n6, res);

		regalloc(&ra, n5.type, N);
		regalloc(&rb, n5.type, N);
		regalloc(&rc, n6.type, N);
		regalloc(&rd, n6.type, N);

		gmove(&n1, &ra);
		gmove(&n3, &rc);
		gins(optoas(OMUL, n5.type), &rc, &ra);	// ra = a*c
		
		gmove(&n2, &rb);
		gmove(&n4, &rd);
		gins(optoas(OMUL, n5.type), &rd, &rb);	// rb = b*d
		gins(optoas(OSUB, n5.type), &rb, &ra);	// ra = (a*c - b*d)

		gins(optoas(OMUL, n5.type), &n2, &rc);	// rc = b*c
		gins(optoas(OMUL, n5.type), &n1, &rd);	// rd = a*d
		gins(optoas(OADD, n5.type), &rd, &rc);	// rc = (b*c + a*d)

		gmove(&ra, &n5);
		gmove(&rc, &n6);

		regfree(&ra);
		regfree(&rb);
		regfree(&rc);
		regfree(&rd);
		break;

	case ODIV:
		if(isimag1i(nr)) {
			complexmove(nl, res, 2);
			break;
		}

		subnode(&n1, &n2, nl);
		subnode(&n3, &n4, nr);
		subnode(&n5, &n6, res);

		regalloc(&ra, n5.type, N);
		regalloc(&rb, n5.type, N);
		regalloc(&rc, n6.type, N);
		regalloc(&rd, n6.type, N);

		gmove(&n1, &ra);
		gmove(&n3, &rc);
		gins(optoas(OMUL, n5.type), &rc, &ra);	// ra = a*c
		
		gmove(&n2, &rb);
		gmove(&n4, &rd);
		gins(optoas(OMUL, n5.type), &rd, &rb);	// rb = b*d
		gins(optoas(OADD, n5.type), &rb, &ra);	// ra = (a*c + b*d)

		gins(optoas(OMUL, n5.type), &n2, &rc);	// rc = b*c
		gins(optoas(OMUL, n5.type), &n1, &rd);	// rd = a*d
		gins(optoas(OSUB, n5.type), &rd, &rc);	// rc = (b*c - a*d)

		gmove(&n3, &rb);
		gins(optoas(OMUL, n5.type), &rb, &rb);	// rb = c*c
		gmove(&n4, &rd);
		gins(optoas(OMUL, n5.type), &rd, &rd);	// rd = d*d
		gins(optoas(OADD, n5.type), &rd, &rb);	// rb = (c*c + d*d)

		gins(optoas(ODIV, n5.type), &rb, &ra);	// ra = (a*c + b*d)/(c*c + d*d)
		gins(optoas(ODIV, n5.type), &rb, &rc);	// rc = (b*c - a*d)/(c*c + d*d)

		gmove(&ra, &n5);
		gmove(&rc, &n6);

		regfree(&ra);
		regfree(&rb);
		regfree(&rc);
		regfree(&rd);
		break;
	}
}

void
complexbool(int op, Node *nl, Node *nr, int true, Prog *to)
{
	Node n1, n2, n3, n4;
	Node na, nb, nc;

	// make both sides addable in ullman order
	if(nr != N) {
		if(nl->ullman > nr->ullman && !nl->addable) {
			tempname(&n1, nl->type);
			complexgen(nl, &n1);
			nl = &n1;
		}
		if(!nr->addable) {
			tempname(&n2, nr->type);
			complexgen(nr, &n2);
			nr = &n2;
		}
	}
	if(!nl->addable) {
		tempname(&n1, nl->type);
		complexgen(nl, &n1);
		nl = &n1;
	}

	// build tree
	// real(l) == real(r) && imag(l) == imag(r)

	subnode(&n1, &n2, nl);
	subnode(&n3, &n4, nr);

	memset(&na, 0, sizeof(na));
	na.op = OANDAND;
	na.left = &nb;
	na.right = &nc;
	na.type = types[TBOOL];

	memset(&nb, 0, sizeof(na));
	nb.op = OEQ;
	nb.left = &n1;
	nb.right = &n3;
	nb.type = types[TBOOL];

	memset(&nc, 0, sizeof(na));
	nc.op = OEQ;
	nc.left = &n2;
	nc.right = &n4;
	nc.type = types[TBOOL];

	if(op == ONE)
		true = !true;

	bgen(&na, true, to);
}

int
cplxsubtype(int et)
{
	if(et == TCOMPLEX64)
		return TFLOAT32;
	if(et == TCOMPLEX128)
		return TFLOAT64;
	fatal("cplxsubtype: %E\n", et);
	return 0;
}

void
nodfconst(Node *n, Type *t, Mpflt* fval)
{
	memset(n, 0, sizeof(*n));
	n->op = OLITERAL;
	n->addable = 1;
	ullmancalc(n);
	n->val.u.fval = fval;
	n->val.ctype = CTFLT;
	n->type = t;

	if(!isfloat[t->etype])
		fatal("nodfconst: bad type %T", t);
}

static int
isimag1i(Node *n)
{
	if(n != N)
	if(n->op == OLITERAL)
	if(n->val.ctype == CTCPLX)
	if(mpgetflt(&n->val.u.cval->real) == 0.0)
	if(mpgetflt(&n->val.u.cval->imag) == 1.0)
		return 1;
	return 0;
}

// break addable nc-complex into nr-real and ni-imaginary
static void
subnode(Node *nr, Node *ni, Node *nc)
{
	int tc;
	Type *t;

	if(!nc->addable)
		fatal("subnode not addable");

	tc = simsimtype(nc->type);
	tc = cplxsubtype(tc);
	t = types[tc];

	if(nc->op == OLITERAL) {
		nodfconst(nr, t, &nc->val.u.cval->real);
		nodfconst(ni, t, &nc->val.u.cval->imag);
		return;
	}

	*nr = *nc;
	nr->type = t;

	*ni = *nc;
	ni->type = t;
	ni->xoffset += t->width;
}

// generate code to negate register nr
static void
negate(Node *nr)
{
	Node nc;
	Mpflt fval;

	memset(&nc, 0, sizeof(nc));
	nc.op = OLITERAL;
	nc.addable = 1;
	ullmancalc(&nc);
	nc.val.u.fval = &fval;
	nc.val.ctype = CTFLT;
	nc.type = nr->type;

	mpmovecflt(nc.val.u.fval, -1.0);
	gins(optoas(OMUL, nr->type), &nc, nr);
}

// generate code to zero addable dest nr
static void
zero(Node *nr)
{
	Node nc;
	Mpflt fval;

	memset(&nc, 0, sizeof(nc));
	nc.op = OLITERAL;
	nc.addable = 1;
	ullmancalc(&nc);
	nc.val.u.fval = &fval;
	nc.val.ctype = CTFLT;
	nc.type = nr->type;

	mpmovecflt(nc.val.u.fval, 0.0);

	gmove(&nc, nr);
}

File: src/cmd/compile/internal/gc/cplx.go
// Copyright 2009 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package gc

import "cmd/internal/obj"

func overlap_cplx(f *Node, t *Node) bool {
	// check whether f and t could be overlapping stack references.
	// not exact, because it's hard to check for the stack register
	// in portable code.  close enough: worst case we will allocate
	// an extra temporary and the registerizer will clean it up.
	return f.Op == OINDREG && t.Op == OINDREG && f.Xoffset+f.Type.Width >= t.Xoffset && t.Xoffset+t.Type.Width >= f.Xoffset
}

func complexbool(op Op, nl, nr, res *Node, wantTrue bool, likely int, to *obj.Prog) {
	// make both sides addable in ullman order
	if nr != nil {
		if nl.Ullman > nr.Ullman && !nl.Addable {
			nl = CgenTemp(nl)
		}

		if !nr.Addable {
			nr = CgenTemp(nr)
		}
	}
	if !nl.Addable {
		nl = CgenTemp(nl)
	}

	// Break nl and nr into real and imaginary components.
	var lreal, limag, rreal, rimag Node
	subnode(&lreal, &limag, nl)
	subnode(&rreal, &rimag, nr)

	// build tree
	// if branching:
	// 	real(l) == real(r) && imag(l) == imag(r)
	// if generating a value, use a branch-free version:
	// 	real(l) == real(r) & imag(l) == imag(r)
	realeq := Node{
		Op:    OEQ,
		Left:  &lreal,
		Right: &rreal,
		Type:  Types[TBOOL],
	}
	imageq := Node{
		Op:    OEQ,
		Left:  &limag,
		Right: &rimag,
		Type:  Types[TBOOL],
	}
	and := Node{
		Op:    OANDAND,
		Left:  &realeq,
		Right: &imageq,
		Type:  Types[TBOOL],
	}

	if res != nil {
		// generating a value
		and.Op = OAND
		if op == ONE {
			and.Op = OOR
			realeq.Op = ONE
			imageq.Op = ONE
		}
		Bvgen(&and, res, true)
		return
	}

	// generating a branch
	if op == ONE {
		wantTrue = !wantTrue
	}

	Bgen(&and, wantTrue, likely, to)
}

// break addable nc-complex into nr-real and ni-imaginary
func subnode(nr *Node, ni *Node, nc *Node) {
	if !nc.Addable {
		Fatalf("subnode not addable")
	}

	tc := Simsimtype(nc.Type)
	tc = cplxsubtype(tc)
	t := Types[tc]

	if nc.Op == OLITERAL {
		u := nc.Val().U.(*Mpcplx)
		nodfconst(nr, t, &u.Real)
		nodfconst(ni, t, &u.Imag)
		return
	}

	*nr = *nc
	nr.Type = t

	*ni = *nc
	ni.Type = t
	ni.Xoffset += t.Width
}

// generate code res = -nl
func minus(nl *Node, res *Node) {
	var ra Node
	ra.Op = OMINUS
	ra.Left = nl
	ra.Type = nl.Type
	Cgen(&ra, res)
}

// build and execute tree
//	real(res) = -real(nl)
//	imag(res) = -imag(nl)
func complexminus(nl *Node, res *Node) {
	var n1 Node
	var n2 Node
	var n5 Node
	var n6 Node

	subnode(&n1, &n2, nl)
	subnode(&n5, &n6, res)

	minus(&n1, &n5)
	minus(&n2, &n6)
}

// build and execute tree
//	real(res) = real(nl) op real(nr)
//	imag(res) = imag(nl) op imag(nr)
func complexadd(op Op, nl *Node, nr *Node, res *Node) {
	var n1 Node
	var n2 Node
	var n3 Node
	var n4 Node
	var n5 Node
	var n6 Node

	subnode(&n1, &n2, nl)
	subnode(&n3, &n4, nr)
	subnode(&n5, &n6, res)

	var ra Node
	ra.Op = op
	ra.Left = &n1
	ra.Right = &n3
	ra.Type = n1.Type
	Cgen(&ra, &n5)

	ra = Node{}
	ra.Op = op
	ra.Left = &n2
	ra.Right = &n4
	ra.Type = n2.Type
	Cgen(&ra, &n6)
}

// build and execute tree
//	tmp       = real(nl)*real(nr) - imag(nl)*imag(nr)
//	imag(res) = real(nl)*imag(nr) + imag(nl)*real(nr)
//	real(res) = tmp
func complexmul(nl *Node, nr *Node, res *Node) {
	var n1 Node
	var n2 Node
	var n3 Node
	var n4 Node
	var n5 Node
	var n6 Node
	var tmp Node

	subnode(&n1, &n2, nl)
	subnode(&n3, &n4, nr)
	subnode(&n5, &n6, res)
	Tempname(&tmp, n5.Type)

	// real part -> tmp
	var rm1 Node

	rm1.Op = OMUL
	rm1.Left = &n1
	rm1.Right = &n3
	rm1.Type = n1.Type

	var rm2 Node
	rm2.Op = OMUL
	rm2.Left = &n2
	rm2.Right = &n4
	rm2.Type = n2.Type

	var ra Node
	ra.Op = OSUB
	ra.Left = &rm1
	ra.Right = &rm2
	ra.Type = rm1.Type
	Cgen(&ra, &tmp)

	// imag part
	rm1 = Node{}

	rm1.Op = OMUL
	rm1.Left = &n1
	rm1.Right = &n4
	rm1.Type = n1.Type

	rm2 = Node{}
	rm2.Op = OMUL
	rm2.Left = &n2
	rm2.Right = &n3
	rm2.Type = n2.Type

	ra = Node{}
	ra.Op = OADD
	ra.Left = &rm1
	ra.Right = &rm2
	ra.Type = rm1.Type
	Cgen(&ra, &n6)

	// tmp ->real part
	Cgen(&tmp, &n5)
}

func nodfconst(n *Node, t *Type, fval *Mpflt) {
	*n = Node{}
	n.Op = OLITERAL
	n.Addable = true
	ullmancalc(n)
	n.SetVal(Val{fval})
	n.Type = t

	if !t.IsFloat() {
		Fatalf("nodfconst: bad type %v", t)
	}
}

func Complexop(n *Node, res *Node) bool {
	if n != nil && n.Type != nil {
		if n.Type.IsComplex() {
			goto maybe
		}
	}

	if res != nil && res.Type != nil {
		if res.Type.IsComplex() {
			goto maybe
		}
	}

	if n.Op == OREAL || n.Op == OIMAG {
		//dump("\ncomplex-yes", n);
		return true
	}

	//dump("\ncomplex-no", n);
	return false

maybe:
	switch n.Op {
	case OCONV, // implemented ops
		OADD,
		OSUB,
		OMUL,
		OMINUS,
		OCOMPLEX,
		OREAL,
		OIMAG:
		//dump("\ncomplex-yes", n);
		return true

	case ODOT,
		ODOTPTR,
		OINDEX,
		OIND,
		ONAME:
		//dump("\ncomplex-yes", n);
		return true
	}

	//dump("\ncomplex-no", n);
	return false
}

func Complexmove(f *Node, t *Node) {
	if Debug['g'] != 0 {
		Dump("\ncomplexmove-f", f)
		Dump("complexmove-t", t)
	}

	if !t.Addable {
		Fatalf("complexmove: to not addable")
	}

	ft := Simsimtype(f.Type)
	tt := Simsimtype(t.Type)
	// complex to complex move/convert.
	// make f addable.
	// also use temporary if possible stack overlap.
	if (ft == TCOMPLEX64 || ft == TCOMPLEX128) && (tt == TCOMPLEX64 || tt == TCOMPLEX128) {
		if !f.Addable || overlap_cplx(f, t) {
			var tmp Node
			Tempname(&tmp, f.Type)
			Complexmove(f, &tmp)
			f = &tmp
		}

		var n1 Node
		var n2 Node
		subnode(&n1, &n2, f)
		var n4 Node
		var n3 Node
		subnode(&n3, &n4, t)

		Cgen(&n1, &n3)
		Cgen(&n2, &n4)
	} else {
		Fatalf("complexmove: unknown conversion: %v -> %v\n", f.Type, t.Type)
	}
}

func Complexgen(n *Node, res *Node) {
	if Debug['g'] != 0 {
		Dump("\ncomplexgen-n", n)
		Dump("complexgen-res", res)
	}

	for n.Op == OCONVNOP {
		n = n.Left
	}

	// pick off float/complex opcodes
	switch n.Op {
	case OCOMPLEX:
		if res.Addable {
			var n1 Node
			var n2 Node
			subnode(&n1, &n2, res)
			var tmp Node
			Tempname(&tmp, n1.Type)
			Cgen(n.Left, &tmp)
			Cgen(n.Right, &n2)
			Cgen(&tmp, &n1)
			return
		}

	case OREAL, OIMAG:
		nl := n.Left
		if !nl.Addable {
			var tmp Node
			Tempname(&tmp, nl.Type)
			Complexgen(nl, &tmp)
			nl = &tmp
		}

		var n1 Node
		var n2 Node
		subnode(&n1, &n2, nl)
		if n.Op == OREAL {
			Cgen(&n1, res)
			return
		}

		Cgen(&n2, res)
		return
	}

	// perform conversion from n to res
	tl := Simsimtype(res.Type)

	tl = cplxsubtype(tl)
	tr := Simsimtype(n.Type)
	tr = cplxsubtype(tr)
	if tl != tr {
		if !n.Addable {
			var n1 Node
			Tempname(&n1, n.Type)
			Complexmove(n, &n1)
			n = &n1
		}

		Complexmove(n, res)
		return
	}

	if !res.Addable {
		var n1 Node
		Igen(res, &n1, nil)
		Cgen(n, &n1)
		Regfree(&n1)
		return
	}

	if n.Addable {
		Complexmove(n, res)
		return
	}

	switch n.Op {
	default:
		Dump("complexgen: unknown op", n)
		Fatalf("complexgen: unknown op %v", n.Op)

	case ODOT,
		ODOTPTR,
		OINDEX,
		OIND,
		OCALLFUNC,
		OCALLMETH,
		OCALLINTER:
		var n1 Node
		Igen(n, &n1, res)

		Complexmove(&n1, res)
		Regfree(&n1)
		return

	case OCONV,
		OADD,
		OSUB,
		OMUL,
		OMINUS,
		OCOMPLEX,
		OREAL,
		OIMAG:
		break
	}

	nl := n.Left
	if nl == nil {
		return
	}
	nr := n.Right

	// make both sides addable in ullman order
	var tnl Node
	if nr != nil {
		if nl.Ullman > nr.Ullman && !nl.Addable {
			Tempname(&tnl, nl.Type)
			Cgen(nl, &tnl)
			nl = &tnl
		}

		if !nr.Addable {
			var tnr Node
			Tempname(&tnr, nr.Type)
			Cgen(nr, &tnr)
			nr = &tnr
		}
	}

	if !nl.Addable {
		Tempname(&tnl, nl.Type)
		Cgen(nl, &tnl)
		nl = &tnl
	}

	switch n.Op {
	default:
		Fatalf("complexgen: unknown op %v", n.Op)

	case OCONV:
		Complexmove(nl, res)

	case OMINUS:
		complexminus(nl, res)

	case OADD, OSUB:
		complexadd(n.Op, nl, nr, res)

	case OMUL:
		complexmul(nl, nr, res)
	}
}
